File: Stack/matchAndBalance.ts

const matchAndBalance = (str: string): boolean => {
    const stack = [];
    const pair = { "}": "{", ")": "(", "]": "[]" }
    for (const char of str) {
        // opeing bracke push in stack
        if (char == "(" || char == "[" || char == "{") {
            stack.push(char)
        }
        else if (char == ")" || char == "]" || char == "}") {
            if (stack.pop() !== pair[char])
                return false
        }
    }

    return stack.length === 0;
}

console.log(matchAndBalance("({{}})"))
console.log(matchAndBalance("}{)(][]"))
console.log(matchAndBalance("{{{{{{{}}}}}}"))

File: Stack/matchAndBalance.py
def matchAndBalance(s: str) -> bool:
    pairs = {
        "}": "{",
        ")": "(",
        "]": "["
    }

    stack = []

    for char in s:

        if char in "({[":
            stack.append(char)

        elif char in ")}]":

            if not stack or stack.pop() != pairs[char]:
                return False

    return len(stack) == 0


print(matchAndBalance("({{}})"))       # True
print(matchAndBalance("}{{{{{"))       # False
print(matchAndBalance("{{{{{{{}}}}}}")) # True